src/TRestGeant4Metadata.cxx

```diff
@@ -936,73 +937,124 @@ void TRestGeant4Metadata::ReadGenerator() {
+        exit(1);
     }
 }
 
+///////////////////////////////////////////////
+/// \brief It reads the <source definition given by argument
+///
+void TRestGeant4Metadata::ReadParticleSource(TRestGeant4ParticleSource* source, TiXmlElement* definition) {
+    TiXmlElement* sourceDefinition = definition;
+
+    source->SetParticleName(GetParameter("particle", sourceDefinition));
+
+    source->SetExcitationLevel(StringToDouble(GetParameter("excitedLevel", sourceDefinition)));
+
+    Int_t charge = 0;
+    if (GetParameter("charge", sourceDefinition) == PARAMETER_NOT_FOUND_STR)
+        charge = 0;
+    else
+        charge = StringToInteger(GetParameter("charge", sourceDefinition));
+
+    source->SetParticleCharge(charge);
+
+    TString fullChain = GetParameter("fullChain", sourceDefinition);
+
+    if (fullChain == "on" || fullChain == "ON" || fullChain == "On" || fullChain == "oN") {
+        SetFullChain(true);
+    } else {
+        SetFullChain(false);
+    }
+
+    // Angular distribution parameters
+    TiXmlElement* angularDefinition = GetElement("angularDist", sourceDefinition);
+    source->SetAngularDistType(GetParameter("type", angularDefinition, "flux"));
+    if (source->GetAngularDistType() == "TH1D") {
+        source->SetAngularFilename(SearchFile(GetParameter("file", angularDefinition)));
+        source->SetAngularName(GetParameter("spctName", angularDefinition));
+    }
+    if (GetNumberOfSources() == 0 && source->GetAngularDistType() == "backtoback") {
+        cout << "WARNING: First source cannot be backtoback. Setting it to isotropic" << endl;
+        source->SetAngularDistType("isotropic");
+    }
+    source->SetDirection(StringTo3DVector(GetParameter("direction", angularDefinition, "(1,0,0)")));
+
+    // Energy distribution parameters
+    TiXmlElement* energyDefinition = GetElement("energyDist", sourceDefinition);
+    source->SetEnergyDistType(GetParameter("type", energyDefinition, "mono"));
+    if (source->GetEnergyDistType() == "TH1D") {
+        source->SetSpectrumFilename(SearchFile(GetParameter("file", energyDefinition)));
+        source->SetSpectrumName(GetParameter("spctName", energyDefinition));
+    }
+    source->SetEnergyRange(Get2DVectorParameterWithUnits("range", energyDefinition));
+    if (source->GetEnergyDistType() == "mono") {
+        Double_t en;
+        en = GetDblParameterWithUnits("energy", energyDefinition, 0);
+        source->SetEnergyRange(TVector2(en, en));
+        source->SetEnergy(en);
+    }
+
+    // allow custom configuration from the class
+    source->LoadConfigFromElement(sourceDefinition, fElementGlobal, fVariables);
+    // AddParticleSource(source);
+}
+
+void TRestGeant4Metadata::RemoveParticleSources() {
+    for (auto c : fParticleSource) {
+        delete c;
+    }
+    fParticleSource.clear();
+}
+
+void TRestGeant4Metadata::AddParticleSource(TRestGeant4ParticleSource* src) {
+    fParticleSource.push_back(src);
```

Comment:
`fParticleSource.emplace_back(src);`

Reply:
What is the difference?

Reply:
`emplace_back` avoids the copy of the element to the vector and thus takes less resources.

https://stackoverflow.com/questions/4303513/push-back-vs-emplace-back

Reply:
So, what happens if the pointer is removed outside or dissapears? It also dissapears inside the vector? That would be a bit unsafe, and need more control on our side.

Reply:
In this case you should do `fParticleSource.emplace_back(std::move(src) );` which still more optimal than `push_back`

Reply:
It is just copying a pointer and I don't think emplace_back saves much resources.